Restore: Associate swap tx IDs from onchain data

ok300 opened the pull request:
This PR recovers the various swap tx IDs from
- initial immutable DB data
- onchain data (LWK)
- a single Electrum batch call to fetch script histories

Open points
- [x] Send swaps
- [x] Receive swaps
- [x] Chain swaps

Fixes #373

roeierez commented on lib/core/src/sdk.rs:
```diff
@@ -1431,6 +1440,118 @@ impl LiquidSdk {
             }
         }
 
+        // TODO Ensure this happens only on first sync
```
Perhaps it is better for now to put all logic of recovering chain data in its own function (recover_swaps_chain_data ?). I am thinking that because it might need to be reusable code and we still not sure how we should incorporate it.

roeierez commented on lib/core/src/sdk.rs:
```diff
@@ -1431,6 +1440,118 @@ impl LiquidSdk {
+                wallet_tx
+                    .tx
+                    .input
+                    .first()
```
I agree it will work but if on future versions we will find a way to aggregate multiple refunds in one transaction (to save fees) it will break. So perhaps just map all inputs?

roeierez reviewed (commented):
Looks like the right direction to me.

ok300 commented:
@roeierez I rewrote the PR to use the script histories fetched from Electrum, as discussed.

This also addressed all your points above.

So input data is now:
- immutable DB data (simulated, I fetch it from the current DB)
- onchain data (our onchain txs from LWK)
- script histories from 1 Electrum call (txs associated with every swap script)

roeierez commented on lib/core/src/restore_onchain.rs:
```diff
@@ -0,0 +1,334 @@
+
+            Ok(())
+        }
+        pub(crate) async fn get_onchain_data(
```
The whole process contains lots of steps and since there are multiple types of data mapping structures  and lots of intermediate steps it makes it hard to follow the code.

In my mind the process could be simplify by following these 4 steps:
1. Fetch wallet transactions, map them by tx id.
2. Fetch two kind of swap items from db (receive, send), map them by script (e.g  Map<script, (swap_id, type)>
3. Fetch script history for all swaps together
4. For each script history:
- Find lockup and claim and add the info to the swap items.

The code roughly does that but I think it can be simplified.
For example we have:
`receive_db: HashMap<String, (CreateReverseResponse, LBtcSwapScript, Script)>`
and 
```
let mut receive_composite_data: HashMap<
                String,
                (CreateReverseResponse, LBtcSwapScript, Script, &Vec<History>)
```
Perhaps it is better and simpler to create a struct with the composite fields ( CreateReverseResponse, LBtcSwapScript, Script, &Vec<History> )
and fill them.
Also instead of &Vec<History> can we populate in that struct directly the lockup & claim ? It will save another processing (mapping) later.

ok300 replied:
Done, added custom types and simplified the various types used.

roeierez commented on lib/core/src/restore_onchain.rs:
```diff
@@ -0,0 +1,334 @@
+
+    use crate::sdk::LiquidSdk;
+
+    pub(crate) struct ImmutableDb {
```
NIT: Wouldn't use immuntable as it doesn't hint anytning about the data it contains. Perhaps just SwapsList ?

roeierez commented on lib/core/src/restore_onchain.rs:
```diff
@@ -0,0 +1,334 @@
+    }
+
+    impl LiquidSdk {
+        pub(crate) async fn get_immutable_db_data(&self) -> Result<ImmutableDb> {
```
NIT: get_swaps_list?

ok300 commented on lib/core/src/chain/liquid.rs:
```diff
@@ -31,9 +31,12 @@ pub trait LiquidChainService: Send + Sync {
     async fn get_script_history(&self, scripts: &Script) -> Result<Vec<History>>;
 
+    /// Get the transactions involved in a list of scripts, including lowball transactions.
+    async fn get_scripts_history_electrum(&self, scripts: &[&Script]) -> Result<Vec<Vec<History>>>;
```
Done: 5eb753846f4a66ee93f4c44fd1b3d78f4ba72c19

ok300 commented on lib/core/src/restore_onchain.rs:
```diff
@@ -0,0 +1,334 @@
+/// Methods to simulate the immutable DB data available from real-time sync
+// TODO Remove once real-time sync is integrated
+pub(crate) mod immutable {
```
Done: 73dfa71e84e7b0f0dd5c08198ffa78ccf52c38e2

ok300 commented on lib/core/src/restore_onchain.rs:
```diff
@@ -0,0 +1,334 @@
+}
+
+/// Methods to restore the swap tx IDs from the onchain data
+pub(crate) mod onchain {
```
Done, flattened it in `restore.rs`: 73dfa71e84e7b0f0dd5c08198ffa78ccf52c38e2

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,820 @@
+
+            match history.btc_claim_script_history.len() {
+                2 => {
+                    // TODO How to tell the claim tx apart from the lockup tx? Is the order in which they're received from Electrum reliable?
```
We might need to fetch these transactions in order to do it reliably.

ok300 replied:
Done, added a call to fetch the BTC txs so we can look at the vouts / vins.

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,820 @@
+
+/// Methods to simulate the immutable DB data available from real-time sync
+// TODO Remove once real-time sync is integrated
+pub(crate) mod immutable {
```
NIT: I don't know if we need another module here. Also immutable is too generic in this context IMO.

ok300 replied:
This module (marked as `TODO Remove once real-time sync is integrated`) is a placeholder for the immutable DB interface.

It contains ways to "simulate" the immutable data (now deriving it from actual DB data) until the interface is available. I placed these helper methods in this module, so it's easy to remove them when the interface is ready.

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,820 @@
+
+    #[allow(dead_code)]
+    #[derive(Clone)]
+    struct SendSwapImmutableData {
```
NIT: also here Immutable seems too generic. Basically all we have here is the swap_script (event he script is derive from it). So perhapas we can rename to SendSwapScriptsData ?

ok300 replied:
I originally started with that, but later realized these structs are inputs from the Immutable DB, representing the immutable data for the swaps. My understanding is that immutable data will be read into such structs, bringing everything that's available (create response, keypairs, etc). Restore is then supposed to use them as inputs.

From this struct, it happens I'm only using the script data for restoring. I named it `SendSwapImmutableData` in the expectation it will include the rest of the immutable data as well.

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,820 @@
+
+impl LiquidSdk {
+    pub(crate) async fn recover_from_onchain(&self, tx_map: TxMap) -> Result<()> {
+        let immutable_db = self.get_swaps_list().await?;
```
NIT: instead of immutable_db perhaps just `swaps_scripts_list`? or `swaps_list`?

roeierez replied:
Also it is better to call this function out of the recover_from_onchain since I can see later that we will need to invoke it on a list that comes from realtime sync (and doesn't exist in the db).

ok300 replied:
Done: da39009268bde34d26b81a505873e8f028acaf0b

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,820 @@
+
+        // TODO Persist updated swaps
+        // TODO     Send updates with found txids
+        // TODO     How to set tx IDs without having to also set state?
```
Why not setting the state? Does anything prevent us from calculating the state based on the onchain data?

ok300 replied:
As discussed, the state is now derived as a "partial state", based on the immutable data and the onchain data.

roeierez reviewed (changes requested):
Looks a lot better. Added some comments.

ok300 commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,820 @@
+            refund_tx_map.len()
+        );
+
+        Ok(RecoveredOnchainDataSend {
```
As discussed, it makes more sense to have the preimage in the immutable data. It can only be retrieved for some states (non-cooperative claim tx), in other cases its only known from interactions with Boltz. So for consistency, it makes more sense to store it in the immutable data, when available (e.g. after claim is sucessful).

ok300 commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,820 @@
+
+        for (swap_id, history) in receive_histories_by_swap_id {
+            match history.len() {
+                2 => {
```
I switch to that logic, ignoring everything that is above 2 txs (address re-use).

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1164 @@
+            (Some(_), None) => PaymentState::Pending,
+            // TODO How to distinguish between Failed and Created (if in both cases, no lockup or claim tx present)
+            //   See https://docs.boltz.exchange/v/api/lifecycle#reverse-submarine-swaps
+            _ => PaymentState::Failed,
```
If no lockup or claim present then it means the payer didn't start paying (swapper didn't broadcast the lockup) from our perspective it is always Created (even on normal flow without restore).

ok300 replied:
No lockup or claim tx can mean both Created (no payment happened yet) or Failed (`invoice.expired`, `swap.expired`, or even `transaction.failed`).

We default to Created now, but I left this comment to point out that we lose this Boltz-specific status information.

roeierez replied:
I see. Wasn't aware that we set it to Failed in these cases. I am not sure BTW we need to set it to Failed in case of receiving but this is another story out of the scope of this issue.

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1164 @@
+                    None => PaymentState::Created,
+                },
+            },
+            None => PaymentState::Created,
```
Same as with send swap this should always be Timeout as on send we automatically attempt to pay and never leave the swap in Created state.
We also ensure the Created ones are converted to Timeout on startup: https://github.com/breez/breez-sdk-liquid/blob/main/lib/core/src/sdk.rs#L184

ok300 replied:
Done: 07759523cb6f57ecee90eaf72a4d9dee591a21ab

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1164 @@
+impl PartialSwapState for RecoveredOnchainDataChainSend {
+    fn get_partial_state(&self) -> PaymentState {
+        // TODO How to detect TimedOut state?
+        //     TimedOut: This covers the case when the swap state is still Created and the swap fails to reach the
```
See comment bellow

ok300 replied:
Done: 07759523cb6f57ecee90eaf72a4d9dee591a21ab

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1164 @@
+        //     Pending state in time. The TimedOut state indicates the lockup tx should never be broadcast.
+        match &self.lbtc_user_lockup_tx_id {
+            Some(_) => match &self.btc_claim_tx_id {
+                Some(_) => PaymentState::Complete,
```
Shouldn't we check if the btc_claim_tx is confirmed or not? If it is still not confirmed should the payment be still in Pending state?

ok300 replied:
I followed the existing state handling logic, where we mark it as Complete as soon as we broadcast the claim:

https://github.com/breez/breez-sdk-liquid/blob/dc0e347e2b1604e8187f30eedc10486323522b91/lib/core/src/chain_swap.rs#L580-L603

roeierez replied:
I see, indeed it is consistent. Question is do we have an issue with chain swap then? @dangeross shouldn't we make the chain swap completed only after our claim tx i s confirmed?

dangeross replied:
It is currently marking the swap complete for both directions when the claim is broadcast. 

Perhaps for the incoming swap we can let the sync mark it complete. For the outgoing swap we are currently not monitoring the BTC address for confirmed transactions (but we could). What would you suggest @roeierez?

roeierez replied:
@dangeross  For send I think we should especially because it is BTC mainnet where transactions may not confirm or go out of the mempool.
Also users see the swap as completed but still don't see the BTC on the destination address.
For receive, you are right the sync can do that same way we do for receive_swap.
But let's not decouple it with this PR. Once we fix it in a following PR we will fix here as well.

ok300 replied:
The chain swap change was done in https://github.com/breez/breez-sdk-liquid/pull/450

The corresponding change to this PR was done in 97d62c5792b7b36da2e99f2926264116ccd59157

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1164 @@
+impl PartialSwapState for RecoveredOnchainDataChainReceive {
+    fn get_partial_state(&self) -> PaymentState {
+        match &self.btc_user_lockup_tx_id {
+            Some(_) => match &self.lbtc_server_claim_tx_id {
```
Do we care about the server claim? As far as I understand it is the last step in the swap which we don't care about. The swap completes when we claim If I get it right.

roeierez replied:
After reading all the code I now see it is not the server claim but is our claim.
So perhaps rename to `lbtc_user_claim_tx_id` ?

ok300 replied:
> After reading all the code I now see it is not the server claim but is our claim. So perhaps rename to `lbtc_user_claim_tx_id` ?

I checked with @dangeross and this is the server claim as per Boltz nomenclature. For them, the initiator (not creator) of the swap is the "user", and txs associated with those funds are user txs. In this case, that's the BTC lockup tx, BTC claim tx and even BTC refund tx.

Conversely, the funds locked by the swapper ("server") are associated with server txs. In this case, that's the LBTC lockup tx and LBTC claim tx (the tx you referred to).

ok300 replied:
> Do we care about the server claim? As far as I understand it is the last step in the swap which we don't care about. The swap completes when we claim If I get it right.

The server claim is our claim, that's the only reason we check this.

ok300 replied:
As discussed in the group chat, this is correct as it is, so marking this as resolved.

roeierez reviewed (changes requested):
Went over the state restoration, now will go over the other parts.

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1172 @@
+    /// reconstructed swap.
+    pub(crate) async fn recover_from_onchain(&self, tx_map: TxMap) -> Result<RecoveredOnchainData> {
+        // Immutable DB, as fetched from endpoint
+        let imm_db = self.get_swaps_list().await?;
```
Ideally this should be given as parameter to this function so we can use it on swaps that are synced in from outside.

ok300 replied:
Done: da39009268bde34d26b81a505873e8f028acaf0b

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1172 @@
+        // Recovered onchain data (txs) per swap
+        let recovered = self.get_onchain_data(tx_map, &imm_db).await?;
+
+        // Validation
+        // Checks if recovered data (txs, partial state) matches known data
+        for send_swap_id in imm_db.send_swap_immutable_db_by_swap_id.keys() {
+            let full_swap = self
+                .persister
+                .fetch_send_swap_by_id(send_swap_id)
+                .ok()
+                .flatten();
+            let recovered_data = recovered.send.get(send_swap_id);
+
+            info!("[Restore Send] Validating {send_swap_id}");
+            match (full_swap, recovered_data) {
+                (Some(expected), Some(recovered)) => {
+                    let exp_lockup_tx_id = expected.lockup_tx_id;
+                    let rec_lockup_tx_id = recovered.lockup_tx_id.as_ref().map(|h| h.txid.to_hex());
+                    info!("lockup_tx_id: {exp_lockup_tx_id:?} / {rec_lockup_tx_id:?}");
+
+                    let exp_refund_tx_id = expected.refund_tx_id;
+                    let rec_refund_tx_id = recovered.refund_tx_id.as_ref().map(|h| h.txid.to_hex());
+                    info!("refund_tx_id: {exp_refund_tx_id:?} / {rec_refund_tx_id:?}");
+
+                    let exp_state = expected.state;
+                    let rec_state = recovered.get_partial_state();
+                    info!("state: {exp_state:?} / {rec_state:?}");
+                }
+                (Some(_), None) => error!("No recovered data for Send Swap {send_swap_id}"),
+                (None, Some(_)) => error!("No expected data for Send Swap {send_swap_id}"),
+                (None, None) => error!("Unknown Send Swap {send_swap_id}"),
+            }
+            info!("[Restore Send] Validated {send_swap_id}");
+        }
```
What is the purpose of this validation step? If everything matches what we already have then why fetch on chain data from the first place? The idea is that we won't have these swaps in the db so we will have nothing to compare to.
Sames feedback goes for the validation step of other swap types.

ok300 replied:
These validations were a way for me to check if the correct tx IDs and states are restored for various swap types.

I removed them: a0c62b802338a258fe48784fffa05b84e402ae3d

I will add real integration tests in a follow-up.

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1172 @@
+        let mut res: HashMap<String, RecoveredOnchainDataReceive> = HashMap::new();
+        for (swap_id, history) in receive_histories_by_swap_id {
+            let (lockup_tx_id, claim_tx_id) = match history.len() {
+                2 => {
```
We should consider here <=2 as there might be only lockup but no claim.

ok300 replied:
Done: c7575afbf96dcf34f1cc40026c7b21241a31c865

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1172 @@
+                    let second = history[1].clone();
+
+                    // If a history tx is a known incoming tx, it's the claim tx
+                    match tx_map.incoming_tx_map.contains_key::<Txid>(&first.txid) {
```
We are missing here a case of refund to the swapper. Case can be that swapper initiated a lockup (first tx), we didn't claim and then swapper refunded itself (in that case the tx doesn't exist on our map).

ok300 replied:
Done: 31ebefe037be9d20133b314a2167b04732b00d2f

roeierez commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1172 @@
+        for (swap_id, history) in chain_receive_histories_by_swap_id {
+            info!("[Recover Chain Receive] Checking swap {swap_id}");
+
+            let (lbtc_server_lockup_tx_id, lbtc_server_claim_tx_id) = match history
```
It is the server lockup and the user claim, isn't it? Shouldn't we rename `lbtc_server_claim_tx_id` to `lbtc_user_claim_tx_id`

ok300 replied:
As discussed elsewhere, they are both "server txs", as they both act on the swapper-locked funds.

ok300 commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1164 @@
+    }
+}
+
+trait PartialSwapState {
```
> Why do we need a trait here compare to only a function for each swap?

I did it that way so the function `get_partial_state` can have the docs in one place, applicable to all instances. IMO it's important to point out why its a partial state and what it means.

> Also IMO it would be more clear to call the function `restore_state()`

Restore implies it's storing it somewhere, which it doesn't. I renamed it to `derive_partial_state`:  9566f7b87390b824b1e2015221b22156eb598054

ok300 commented on lib/core/src/restore.rs:
```diff
@@ -0,0 +1,1055 @@
+                1 => (Some(history.btc_lockup_script_history[0].clone()), None),
+
+                // Both txs available (lockup + claim, or lockup + refund)
+                // Any tx above the first two, we ignore, as that is address re-use which is not supported
```
Which lines do you mean? In L436-462 it's already checking for `n >= 2` txs and ignores the ones after the 2nd.

dangeross replied:
Ah yeah right, I missed that sorry

dangeross commented:
Just needs conflicts resolving
